3_harmonize/src/clean_wqp_data.R

```diff
@@ -3,29 +3,103 @@
 
   # Clean data and assign flags if applicable
   wqp_data_clean <- wqp_data %>%
     # harmonize characteristic names by assigning a common parameter name
     # to the groups of characteristics supplied in `char_names_crosswalk`.
-    left_join(y = char_names_crosswalk, by = c("CharacteristicName" = "char_name"))
+    left_join(y = char_names_crosswalk, by = c("CharacteristicName" = "char_name")) %>%
+    # flag true missing results, i.e. when result measure value and detection
+    # limit value are both NA, when "not reported" is found in the column
+    # "ResultDetectionConditionText", or when any of the strings from
+    # `commenttext_missing` are found in the column "ResultCommentText".
+    mutate(flag_missing_result = case_when(
+      is.na(ResultMeasureValue) & is.na(DetectionQuantitationLimitMeasure.MeasureValue) ~ TRUE,
+      grepl("not reported", ResultDetectionConditionText, ignore.case = TRUE) ~ TRUE,
+      grepl(paste(commenttext_missing, collapse = "|"), ResultCommentText, ignore.case = TRUE) ~ TRUE
+    )) %>%
+    # Flag duplicate records
+    flag_duplicates(., duplicate_definition)
   
   return(wqp_data_clean)
   
   
 }
 
+
+#' @title Flag duplicated records
+#' 
+#' @description 
+#' Function to flag duplicated rows based on a user-supplied definition
+#' of a duplicate record. Additional flags can help reconcile duplicate sets.
+#' For example, a flag "flag_duplicate_drop_random" is assigned based on 
+#' randomly selecting the first record from a duplicated set, making it 
+#' easier to drop all others from the dataset. 
+#' 
+#' @param wqp_data data frame containing the data downloaded from the WQP, 
+#' where each row represents a unique data record.
+#' @param duplicate_definition character string(s) indicating which columns are
+#' used to identify a duplicate record. Duplicate records are defined as those 
+#' that share the same value for each column within `duplicate_definition`.
+#'
+#' @returns 
+#' Returns a data frame containing data downloaded from the Water Quality Portal,
+#' where each row represents a unique data record. New columns appended to the 
+#' original data frame include flags for duplicated records. 
+#' 
+flag_duplicates <- function(wqp_data, duplicate_definition){
+  
+  wqp_data_out <- wqp_data %>%
+    # Step 1: Flag duplicate records using the `duplicate_definition`
+    group_by(across(all_of(duplicate_definition))) %>% 
+    mutate(n_duplicated = n(),
+           flag_duplicated_row = if_else(n_duplicated > 1, TRUE, NA)) %>% 
+    # Step 2: For remaining duplicates, randomly select the first record from
+    # each duplicated set and flag all others for exclusion
+    mutate(dup_number = seq(n_duplicated),
+           flag_duplicate_drop_random = if_else(n_duplicated > 1 & 
+                                                  dup_number != 1, 
+                                                TRUE, NA)) %>%
```

Comment:
Both here and for `flag_missing_result`, I would expect the possible values to be TRUE and FALSE, not `NA`. What is your reasoning behind the NA?

Reply:
I don't really have a good reason...my one little complaint about `case_when` is that to use `FALSE` for `flag_missing_result` I would type the following:

```r
    mutate(flag_missing_result = case_when(
      is.na(ResultMeasureValue) & is.na(DetectionQuantitationLimitMeasure.MeasureValue) ~ TRUE,
      grepl("not reported", ResultDetectionConditionText, ignore.case = TRUE) ~ TRUE,
      grepl(paste(commenttext_missing, collapse = "|"), ResultCommentText, ignore.case = TRUE) ~ TRUE,
      TRUE ~ FALSE
    ))
```

So that's why I used `NA` but I think the output is more intelligible using the logical values `TRUE` and `FALSE` so will revert to that.

Reply:
I agree, the default for `case_when()` is not very user friendly. I suppose the alternative is that you actually don't need `case_when()`, right? All of the statements are already returning a logical. I think you could just separate each with an `|` "OR" operator.

```r
    mutate(flag_missing_result = 
      ( is.na(ResultMeasureValue) & is.na(DetectionQuantitationLimitMeasure.MeasureValue) ) |
      grepl("not reported", ResultDetectionConditionText, ignore.case = TRUE) |
      grepl(paste(commenttext_missing, collapse = "|"), ResultCommentText, ignore.case = TRUE)
    )
```

Reply:
Realizing I used `case_when` because I thought the alternative was hard to read, and then I complained about `case_when` being hard to read.. 🙃 Maybe it's the length of the WQP column names that gets me!

I like your suggested change and will revert to just using the OR operator.